dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -217,15 +228,49 @@ public void start(WorkerNetAddress address) throws IOException {
     // the heartbeat is only used to notify the aliveness of this worker, so that clients
     // can get the latest worker list from master.
     // TODO(bowen): once we set up a worker discovery service in place of master, remove this
-    getExecutorService()
-        .submit(new HeartbeatThread(HeartbeatContext.WORKER_BLOCK_SYNC,
-            mResourceCloser.register(new BlockMasterSync()),
-            () -> new FixedIntervalSupplier(Configuration.getMs(
-                PropertyKey.WORKER_BLOCK_HEARTBEAT_INTERVAL_MS)),
-            mConf, ServerUserState.global()));
+    // TODO(lucy): temporary fallback logic during transition of removing master dependency
+    if (mMembershipManager instanceof NoOpMembershipManager) {
+      getExecutorService()
+          .submit(new HeartbeatThread(HeartbeatContext.WORKER_BLOCK_SYNC,
+              mResourceCloser.register(new BlockMasterSync()),
+              () -> new FixedIntervalSupplier(Configuration.getMs(
+                  PropertyKey.WORKER_BLOCK_HEARTBEAT_INTERVAL_MS)),
+              mConf, ServerUserState.global()));
+    }
   }
 
+  /**
+   * Register to join to the distributed membership.
+   * @throws IOException
+   */
   private void register() throws IOException {
+    Preconditions.checkState(mAddress != null, "worker not started");
+    RetryPolicy retry = RetryUtils.defaultWorkerMasterClientRetry();
+    // For regression purpose, use the original way of regsiter
+    if (mMembershipManager instanceof NoOpMembershipManager) {
+      registerToMaster();
+      return;
+    }
+    while (true) {
```

Comment:
nit: a little log here, as you might get stuck in this loop?

Reply:
RetryPolicy has a max duration of attempted retry.